### PR TITLE
Initial work on integrating "disjoint group" RoutingTable.

### PR DESCRIPTION
**Pull request by Fraser999**

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
r? @afck

(maidsafe_highfive has picked a reviewer for you, use r? to override)

File: src/core_tests.rs [hunk 1]
```diff
@@ -398,28 +396,30 @@ fn random_churn<R: Rng>(rng: &mut R,
...
+//                           -> BTreeSet<XorName> {
+//     let our_name = table.our_name();
+//     let far_name = our_name.with_flipped_bit(bucket_index);
+
+//     table.closest_nodes_to(&far_name, MIN_GROUP_SIZE, false)
+//         .into_iter()
+//         .map(|info| *info.name())
+//         .filter(|name| our_name.bucket_index(name) == bucket_index)
+//         .collect()
+// }
```
Comment (dirvine) [inline src/core_tests.rs]: Do we use these at all ? If not I would remove them, otherwise they should be completed. Is the intention to do that in a separate PR or remove them altogether

File: src/core_tests.rs [hunk 2]
```diff
@@ -580,12 +580,12 @@ fn less_than_group_size_nodes() {
 
 #[test]
 fn group_size_nodes() {
-    test_nodes(GROUP_SIZE);
+    test_nodes(MIN_GROUP_SIZE);
```
Comment (dirvine) [inline src/core_tests.rs]: I think MIN is implicit

File: src/core_tests.rs [hunk 3]
```diff
@@ -429,37 +429,37 @@ fn sort_nodes_by_distance_to(nodes: &mut [TestNode], name: &XorName) {
...
-                    target,
-                    expected_bucket,
-                    actual_bucket);
-        }
-        count -= expected_bucket.len();
-        bucket_index += 1;
-    }
+/// Verify that the invariant is upheld for the routing table at `index`.
+pub fn verify_invariant(_routing_tables: &[RoutingTable<XorName>], _index: usize) {
+    // let target = routing_tables[index].our_name();
```
Comment (dirvine) [inline src/core_tests.rs]: Could this not be that our group is => GROUP_SIZE and our neighbor group. I wonder do we need more than that? Be good to understand the invariant here.

File: src/error.rs [hunk 4]
```diff
@@ -70,18 +71,12 @@ pub enum RoutingError {
...
-    /// Routing table did not add the node information, either because it was already added, or
-    /// because it did not improve the routing table
-    RefusedFromRoutingTable,
-    /// Rejected providing the close group, because the destination address does not match any of
-    /// the sender's buckets
-    RejectedGetCloseGroup,
     /// A client with `client_restriction == true` tried to send a message restricted to nodes.
     RejectedClientMessage,
+    /// Routing Table error
+    RoutingTable(RoutingTableError),
```
Comment (dirvine) [inline src/error.rs]: Nice move.

File: src/states/node.rs [hunk 5]
```diff
@@ -1272,7 +1260,7 @@ impl Node {
             return Ok(());
         }
 
-        let public_ids = match self.peer_mgr.routing_table().close_nodes(&dst_name, GROUP_SIZE) {
+        let public_ids = match self.peer_mgr.routing_table().close_names(&dst_name) {
             Some(close_group) => self.peer_mgr.get_pub_ids(&close_group),
```
Comment (afck) [inline src/states/node.rs]: Possibly `get_pub_ids` should even be generic in its return type, to avoid the `into_iter().collect()` below.

File: src/states/node.rs [hunk 6]
```diff
@@ -1568,25 +1511,20 @@ impl Node {
...
                    sent_to: &[XorName])
                    -> Result<(Vec<XorName>, Vec<PeerId>), RoutingError> {
         if self.is_proper() {
-            let destination = routing_msg.dst.to_destination();
-            let targets = self.peer_mgr
+            let targets = try!(self.peer_mgr
                 .routing_table()
-                .target_nodes(destination, hop, route as usize)
-                .into_iter()
-                .filter(|target| !sent_to.contains(target.name()))
```
Comment (afck) [inline src/states/node.rs]: If the `sent_to` filtering isn't needed anymore, that should probably be removed everywhere.

File: src/states/node.rs [hunk 7]
```diff
@@ -1706,41 +1645,33 @@ impl Node {
...
-            if let Err(e) = self.request_bucket_ids(bucket_index) {
-                debug!("{:?} Failed to request replacement connection_info from bucket \
-                        {}: {:?}.",
-                       self,
-                       bucket_index,
-                       e);
-            }
+        if let RemovalDetails { targets_and_merge_details: Some((_targets, _merge_details)), .. } =
+               details {
+            // notify peers
```
Comment (afck) [inline src/states/node.rs]: That should probably be a `TODO`.

File: src/utils.rs [hunk 8]
```diff
@@ -127,27 +126,27 @@ mod tests {
         let second_closest = close_nodes[1];
         let mut combined: Vec<u8> = Vec::new();
 
-        for i in original_name.0.into_iter() {
+        for i in &original_name.0 {
```
Comment (afck) [inline src/utils.rs]: Why the `&`? It should be okay to consume that collection.

File: src/states/node.rs [hunk 9]
```diff
@@ -856,78 +851,72 @@ impl Node {
...
                     if let Err(err) = self.event_sender.send(event) {
                         error!("{:?} Error sending event to routing user - {:?}", self, err);
                     }
                 }
             }
         }
 
-        if self.peer_mgr.routing_table().len() == 1 {
-            self.request_bucket_close_groups();
-        }
```
Comment (afck) [inline src/states/node.rs]: With the bucket requests gone, we currently rely exclusively on the peers handling the `NewNode` messages, and we don't ever fix the network if that is missed (e.g. because two nodes join at the same time), I think. That's fine for now, but we need to keep in mind that this is an issue: The consensus handling for the group member lists will need to make sure that all members of our own group _and_ all neighbouring groups eventually agree on the group constituency.

File: src/utils.rs (hunk 8 above)
Comment (fizyk20) [inline src/utils.rs]: It's the same thing we were just discussing - IntoIterator being only implemented for slices and not for arrays.

File: src/states/node.rs (hunk 6 above)
Comment (Fraser999) [inline src/states/node.rs]: I thought it might be useful for debugging for a while, but I'm happy to remove it.  I'll leave it until you confirm one way or the other if you don't mind?

File: src/states/node.rs (hunk 7 above)
Comment (Fraser999) [inline src/states/node.rs]: True.  I'll make the change.

File: src/core_tests.rs (hunk 1 above)
Comment (afck) [inline src/core_tests.rs]: OK, you're right. :+1:

File: src/core_tests.rs [hunk 10]
```diff
@@ -602,17 +602,17 @@ fn failing_connections_group_of_three() {
...
     drop_node(&mut nodes, 1); // Drop a tunnel client. The others should be notified.
-    verify_kademlia_invariant_for_all_nodes(&nodes);
+    verify_invariant_for_all_nodes(&nodes);
 }
 
 #[test]
 fn failing_connections_ring() {
     let network = Network::new(None);
-    let len = GROUP_SIZE * 2;
+    let len = MIN_GROUP_SIZE * 2;
```
Comment (afck) [inline src/core_tests.rs]: :+1:

File: src/peer_manager.rs [hunk 11]
```diff
@@ -562,15 +564,15 @@ impl PeerManager {
...
+    pub fn get_peer_ids(&self, names: &HashSet<XorName>) -> Vec<PeerId> {
         names.iter()
             .filter_map(|name| self.peer_map.get_by_name(name).and_then(Peer::peer_id))
             .cloned()
             .collect()
     }
 
     /// Return the PublicIds of nodes bearing the names.
-    pub fn get_pub_ids(&self, names: &[XorName]) -> Vec<PublicId> {
+    pub fn get_pub_ids(&self, names: &HashSet<XorName>) -> HashSet<PublicId> {
```
Comment (afck) [inline src/peer_manager.rs]: I think the `names.contains()` call isn't really natural: Shouldn't that just be checked in the `else` inside the iteration?

Anyway, I don't feel strongly about that one. Happy to leave it as is.

File: src/states/client.rs [hunk 12]
```diff
@@ -183,7 +182,7 @@ impl Client {
         try!(signed_msg.check_integrity());
 
         // Prevents someone sending messages repeatedly to us
-        if self.signed_msg_filter.filter_incoming(signed_msg) > GROUP_SIZE {
+        if self.signed_msg_filter.filter_incoming(signed_msg) > MIN_GROUP_SIZE {
```
Comment (afck) [inline src/states/client.rs]: :+1:

File: src/states/node.rs [hunk 13]
```diff
@@ -856,78 +851,72 @@ impl Node {
...
                 }
 
-                self.reset_bucket_refresh_timer();
+                let all_rt_contacts = self.peer_mgr
+                    .routing_table()
+                    .iter()
+                    .filter(|name| *name != public_id.name())
+                    .cloned()
+                    .collect();
+                let peer_ids_to_notify = self.peer_mgr.get_peer_ids(&all_rt_contacts);
```
Comment (afck) [inline src/states/node.rs]: I'm slightly in favour of changing it, but I don't insist. :-)

And the borrow checker should also be happy, because neither call is `mut`.

File: src/states/node.rs (hunk 5 above)
Comment (afck) [inline src/states/node.rs]: That's also an option.
And I'm happy with having it in a separate PR.

File: src/states/node.rs (hunk 6 above)
Comment (afck) [inline src/states/node.rs]: :+1: 
But maybe add a TODO about it (or keep a note) somewhere.

File: src/peer_manager.rs (hunk 11 above)
Comment (Fraser999) [inline src/peer_manager.rs]: Well, it's probably not something I'd worry about too much until profiling, but checking in the `else` wouldn't be quite as efficient as calling `contains()` I guess - but clarity would trump performance for me, so yes, I'd say it's clearer to just do the additional check for our own name in the existing `filter_map()` run.

Anyway, I'll leave as is for now to allow us to progress this PR.
